dipy/align/imaffine.py

```diff
@@ -0,0 +1,1010 @@
+            if apply_inverse:
+                sampling_grid2world = self.codomain_grid2world
+            else:
+                sampling_grid2world = self.domain_grid2world
+        if sampling_grid2world is None:
+            sampling_grid2world = np.eye(dim + 1)
+
+        # Obtain world-to-grid transform for input image
+        if image_grid2world is None:
+            if apply_inverse:
+                image_grid2world = self.domain_grid2world
+            else:
+                image_grid2world = self.codomain_grid2world
+            if image_grid2world is None:
+                image_grid2world = np.eye(dim + 1)
+        image_world2grid = npl.inv(image_grid2world)
+
+        # Obtain the appropriate transform method
+        if interp == 'nearest':
+            # Convert input image dtype to an appropriate supported type
+            # Integral supported types are int16 and int32
+            if image.dtype in [np.dtype('float64'), np.dtype('float32')]:
+                image = image.astype(floating)
+            elif image.dtype is np.dtype('int64'):
+                image = image.astype(np.int32)
+            elif image.dtype is np.dtype('int8'):
+                image = image.astype(np.int16)
+            if dim == 2:
+                transform_method = vf.warp_2d_affine_nn
+            elif dim == 3:
+                transform_method = vf.warp_3d_affine_nn
+        elif interp == 'linear':
+            # for trilinear interpolation, the only supported type is floating
+            image = image.astype(floating)
+            if dim == 2:
+                transform_method = vf.warp_2d_affine
+            elif dim == 3:
+                transform_method = vf.warp_3d_affine
+        else:
+            raise ValueError('Unknown interpolation method: '+str(interp))
+
+        # Compute the transform from sampling grid to input image grid
+        if apply_inverse:
+            aff = self.affine_inverse
+        else:
+            aff = self.affine
+
+        if (aff is None) or resample_only:
+            comp = image_world2grid.dot(sampling_grid2world)
+        else:
+            comp = image_world2grid.dot(self.affine.dot(sampling_grid2world))
+
+        # Transform the input image
+        transformed = transform_method(image, shape, comp)
+        return transformed
+
+    def transform(self, image, interp='linear', image_grid2world=None,
+                  sampling_grid_shape=None, sampling_grid2world=None,
+                  resample_only=False):
+        """ Transforms the input image from co-domain to domain space
+
+        By default, the transformed image is sampled at a grid defined by
+        `self.domain_shape` and `self.domain_grid2world`. If such
+        information was not provided then `sampling_grid_shape` is mandatory.
+
+        Parameters
+        ----------
+        image : array, shape (X, Y) or (X, Y, Z)
+            the image to be transformed
+        interp : string, either 'linear' or 'nearest'
```

Comment:
>I can easily imagine wanting to do spline interpolation here. Is that hard to implement?

No, it's not hard. We just need to define new warping functions like those in `vector_fields`. The problem is just cython limitations. Maybe the warping functions should receive the pointer to the interpolation function to be used to warp the full volume. The problem with this would be the overhead of calling the interpolation function for each voxel. Right now, the functions that interpolate the volume at a single point are `inline`, I remember that before defining them as `inline` the diffeomorphic registration was significantly slower, although I don't remember exactly how much.